lightning/src/ln/peer_handler.rs

```diff
@@ -1272,4 +1284,124 @@ mod tests {
+		let node_2_btckey = SecretKey::from_slice(&[39; 32]).unwrap();
+		let unsigned_ann = msgs::UnsignedChannelAnnouncement {
+			features: ChannelFeatures::supported(),
+			chain_hash: genesis_block(network).header.bitcoin_hash(),
+			short_channel_id: short_chan_id,
+			node_id_1: PublicKey::from_secret_key(&secp_ctx, &node_1_privkey),
+			node_id_2: PublicKey::from_secret_key(&secp_ctx, &node_2_privkey),
+			bitcoin_key_1: PublicKey::from_secret_key(&secp_ctx, &node_1_btckey),
+			bitcoin_key_2: PublicKey::from_secret_key(&secp_ctx, &node_2_btckey),
+			excess_data: Vec::new(),
+		};
+
+		let msghash = Message::from_slice(&Sha256dHash::hash(&unsigned_ann.encode()[..])[..]).unwrap();
+		let node_1_sig = secp_ctx.sign(&msghash, &node_1_privkey);
+		let node_2_sig = secp_ctx.sign(&msghash, &node_2_privkey);
+		let bitcoin_1_sig = secp_ctx.sign(&msghash, &node_1_btckey);
+		let bitcoin_2_sig = secp_ctx.sign(&msghash, &node_2_btckey);
+
+		msgs::ChannelAnnouncement {
+			node_signature_1: node_1_sig,
+			node_signature_2: node_2_sig,
+			bitcoin_signature_1: bitcoin_1_sig,
+			bitcoin_signature_2: bitcoin_2_sig,
+			contents: unsigned_ann,
+		}
+	}
+
+	fn get_dummy_channel_update(short_chan_id: u64) -> msgs::ChannelUpdate {
+		use secp256k1::ffi::Signature as FFISignature;
+		let network = Network::Testnet;
+		msgs::ChannelUpdate {
+			signature: Signature::from(FFISignature::new()),
+			contents: msgs::UnsignedChannelUpdate {
+				chain_hash: genesis_block(network).header.bitcoin_hash(),
+				short_channel_id: short_chan_id,
+				timestamp: 0,
+				flags: 0,
+				cltv_expiry_delta: 0,
+				htlc_minimum_msat: 0,
+				fee_base_msat: 0,
+				fee_proportional_millionths: 0,
+				excess_data: vec![],
+			}
+		}
+	}
+
+	#[test]
+	fn test_do_attempt_write_data() {
+		// Create 2 peers with custom TestRoutingMessageHandlers and connect them.
+		let chan_handlers = create_chan_handlers(2);
+		let mut routing_handlers: Vec<Arc<msgs::RoutingMessageHandler>> = Vec::new();
+		for i in 0..2 {
+			let routing_handler_option1 = TestRoutingMessageHandler::new();
+			let routing_handler_option2 = TestRoutingMessageHandler::new();
+			if i == 0 {
+				routing_handlers.push(Arc::new(routing_handler_option1));
+			} else {
+				routing_handlers.push(Arc::new(routing_handler_option2));
+			}
+		}
+		let peers = create_network(2, &chan_handlers, Some(&routing_handlers));
+
+		// By calling establish_connect, we trigger do_attempt_write_data between
+		// the peers. Previously this function would mistakenly enter an infinite loop
+		// when there were more channel messages available than could fit into a peer's
+		// buffer. This issue would now be detected by this test (because we use custom
+		// RoutingMessageHandlers that intentionally return more channel messages
+		// than can fit into a peer's buffer).
+		establish_connection(&peers[0], &peers[1]);
+		assert_eq!(peers[0].peers.lock().unwrap().peers.len(), 1);
```

Comment:
Maybe also assert that N channel_update/announcements have been received by each peer (by tracking how many are received in TestRoutingMessageHandlers and keeping an Arc<> copy to both of them outside the Vec?)

Reply:
Good idea. Still figuring out why one of them receives one less channel announcement than the other so I can add a comment

Reply:
Matt, just checking - lmk if adding explanation comments for the # of updates received by each peer is a blocker. If not, all comments are addressed.

Reply:
Oops, sorry, based on your comment I figured you were still working on this. What *is* the reason for this? It does seem like a bug.

Reply:
They receive the same number of updates when [`end`](https://github.com/rust-bitcoin/rust-lightning/pull/556/files#diff-009d5545bd3452d77ff37b281993ced7R1317) is hard-coded. So it's def related to each of the peer's [`pending_outbound_buffer`s](https://github.com/rust-bitcoin/rust-lightning/pull/556/files#diff-009d5545bd3452d77ff37b281993ced7R356) being populated differently (probably in response to the messages from the remote peer?). 

Since it seems like a bug, I'll look into why there's a difference

Reply:
Ok, it's just because peer_a already has an Init message in its buffer of messages to send to peer_b, which affects the calculation of batch_amount for `get_next_channel_announcements`. Added a comment!

Reply:
Ah, right. So its not a bug, but it does sound like a fragile test. Maybe change end to be some calculation on chan_anns_recvd so that we always get to some expected total?

Reply:
Might be misunderstanding you but that sounds like a circular dependency since `chan_anns_recvd` inherently depends on `end`'s value, which inherently depends on `batch_amount`, which is fragile...?

How about this update (I just raised the number of channels to sync to 100)? Is there a risk of `batch_amount` being raised significantly?

Reply:
Or, could have it depend on MSG_BUFFER_SIZE, which'll always be greater than `batch_amount`..

Reply:
Errr, I meant to just call end min(batch_amount, CONST_VALUE - other_peer.chan_anns_recvd.load())

Reply:
Sorry but why subtract `other_peer.chan_anns_recvd.load()`? (& I assume you mean `min(start + batch_amount, ...)` ?

Reply:
Well what matters is number of updates *sent*, hence I mentioned other_peer here. And, yea, + start.

Reply:
Thanks, updated!